Added preprocess option to support timeSeriesStats times using daysSinceStartOfSim

milenaveneziani opened the pull request:
@pwolfram: here are my changes for supporting the time of timeSeriesStats AM fields, using the daysSinceStartOfSim variable as we discussed. I added to the test case (and included instructions in the README file) so that a short time series of global ice concentration can be plotted with the preprocess based on xtime and the preprocess based on daysSinceStartOfSim.

Besides the test case, here is what I get from one of the ACME coupled runs if I use the simple preprocess_mpas:
![screen shot 2016-04-19 at 1 33 48 am](https://cloud.githubusercontent.com/assets/8237579/14631156/491c200e-05cf-11e6-9578-0bb6b6b086e5.png)

and instead here is the result if I use preprocess_mpas_timeSeriesStats:

![screen shot 2016-04-19 at 1 34 14 am](https://cloud.githubusercontent.com/assets/8237579/14631166/5b7d656e-05cf-11e6-8d80-835ff7fbb5ea.png)

pwolfram commented:
@milenaveneziani, this looks mostly good to me (although I have a few items I'll mark in the code).  I have also submitted a PR to your branch with some general formatting cleanup.  The only substantial thing I have is for `test_load_mpas_xarray_timeSeriesStats_datasets`.  I think the plot produced 

![screenshot 2016-04-19 08 49 15](https://cloud.githubusercontent.com/assets/4295853/14643462/b9a7f82a-060b-11e6-8788-ac949143cad0.png)

could be a little clearer because it is not immediate obviously to me that the new functionality meets the design criteria stated in https://github.com/pwolfram/mpas_xarray_wrapper/issues/4.  For instance, it seems like an ideal example would start on Jan 1 1850.  I think this is correct based on what I can tell, e.g.,

``` bash
┌─[pwolfram][shapiro][~/src/mpas_xray_wrapper][08:52][±][pwolfram/preprocess_timeserisstats ✓]
└─▪ ncdump -v timeSeriesStatsMonthly_avg_daysSinceStartOfSim_1 am.mpas-cice.timeSeriesStatsMonthly.0001.nc  | tail
        :config_AM_timeSeriesStatsClimatology_duration_intervals = "00-03-00_00:00:00;00-03-00_00:00:00;00-03-00_00:00:00;00-03-00_00:00:00;00-03-00_00:00:00;00-03-00_00:00:00;00-03-00_00:00:00;00-03-00_00:00:00" ;
        :config_AM_timeSeriesStatsClimatology_repeat_intervals = "01-00-00_00:00:00;01-00-00_00:00:00;01-00-00_00:00:00;01-00-00_00:00:00;01-00-00_00:00:00;01-00-00_00:00:00;01-00-00_00:00:00;01-00-00_00:00:00" ;
        :config_AM_timeSeriesStatsClimatology_reset_intervals = "10-00-00_00:00:00;10-00-00_00:00:00;10-00-00_00:00:00;10-00-00_00:00:00;10-00-00_00:00:00;10-00-00_00:00:00;10-00-00_00:00:00;10-00-00_00:00:00" ;
        :file_id = "4dx1ado52r" ;
        :NCO = "\"4.5.6-alpha03\"" ;
data:

 timeSeriesStatsMonthly_avg_daysSinceStartOfSim_1 = 288.306117955832, 
    318.778474906597 ;
}

```

``` python
In [1]: import datetime

In [2]: datetime.datetime(1849,12,31) + datetime.timedelta(288.306117955832)
Out[2]: datetime.datetime(1850, 10, 15, 7, 20, 48, 591384)

```

However, if it is easier to make a cleaner example, ideally starting a few days into Jan 1850, we should but otherwise this appears to work as far as I can tell and I'm happy to merge.

pwolfram commented on mpas_xarray.py:
```diff
@@ -54,6 +54,34 @@ def preprocess_mpas(ds, yearoffset=1700): #{{{
 
     return ds #}}}
 
+def preprocess_mpas_timeSeriesStats(ds, yearoffset=1849): #{{{
```
We should specify in the comments why year 1849 was chosen as the year offset.

pwolfram commented on mpas_xarray.py:
```diff
@@ -54,6 +54,34 @@ def preprocess_mpas(ds, yearoffset=1700): #{{{
+    """
+
+    daysSinceStart = ds.timeSeriesStatsMonthly_avg_daysSinceStartOfSim_1
+    time = [datetime.datetime(yearoffset,12,31) + datetime.timedelta(x) for x in daysSinceStart.values]
```
May also want to include month and day offsets as arguments to `preprocess_mpas_timeSeriesStats` above.  I think this would be helpful for generality and to avoid "magic" numbers.

pwolfram commented on mpas_xarray.py:
```diff
@@ -28,7 +28,7 @@
 import xarray as xr
 
-def preprocess_mpas(ds, yearoffset=1700): #{{{
+def preprocess_mpas(ds, yearoffset=1849): #{{{
```
Need a comment in the doc string as to why this is the default yearoffset because there if physical meaning to the choice of 1849.

pwolfram commented on mpas_xarray.py:
```diff
@@ -54,6 +54,34 @@ def preprocess_mpas(ds, yearoffset=1700): #{{{
+    assert datetimes[0].year > 1678, 'ERROR: yearoffset=%s'%(yearoffset) + \
+            ' must be large enough to ensure datetimes larger than year 1678'
+    assert datetimes[-1].year < 2262, 'ERROR: yearoffset=%s'%(yearoffset) + \
+            ' must be large enough to ensure datetimes larger than year 2262'
```
This code is duplicated with the code above in the function.  Probably this set of asserts should be written as a function for cleaner code.

pwolfram commented on README.md:
```diff
@@ -5,7 +5,10 @@ Wrapper to handle importing MPAS files into xarray (https://github.com/pydata/xa
- 2. Provides capability to remove redundant time entries from
+ 2. Converts MPAS "timeSinceStartOfSim" to xarray time optionally for MPAS fields coming from the
+    timeSeriesStatsAM.  Time dimension is assigned via 
+    `preprocess_mpas_timeSeriesStats`.
```
I don't think this has to be on its own line.  I would place line 10's contents on line 9.

pwolfram commented on README.md:
```diff
@@ -5,7 +5,10 @@ Wrapper to handle importing MPAS files into xarray (https://github.com/pydata/xa
  1. Converts MPAS "xtime" to xarray time.  Time dimension is assigned via
     `preprocess_mpas`.
- 2. Provides capability to remove redundant time entries from
+ 2. Converts MPAS "timeSinceStartOfSim" to xarray time optionally for MPAS fields coming from the
```
I don't think we need the word optionally

pwolfram commented on README.md:
```diff
@@ -27,3 +30,23 @@ python mpas_xarray.py -f "globalStats*nc"
+
+This plots a short time series of global average ice concentration, showing the correctly centered curve
+(derived using preprocess_mpas_timeSeriesStats) and the curve incorrectly shifted toward the end of the 
+time averaging period (derived using preprocess_mpas).
```
We could include an image for the examples if you think this is helpful.  However, it is not necessary and may not be useful.  I could go either way on this.

pwolfram commented on README.md:
```diff
@@ -27,3 +30,23 @@ python mpas_xarray.py -f "globalStats*nc"
+This plots a short time series of global average ice concentration, showing the correctly centered curve
+(derived using preprocess_mpas_timeSeriesStats) and the curve incorrectly shifted toward the end of the 
+time averaging period (derived using preprocess_mpas).
+
```
Probably don't need trailing whitespace line.

pwolfram commented on mpas_xarray.py:
```diff
@@ -91,16 +119,40 @@ def test_load_mpas_xarray_datasets(path):
+    var_ice  = var.where(ind_ice)
+    var_ice =  var_ice.mean('nCells')
+    var2_ice = var2.where(ind_ice2)
+    var2_ice = var2_ice.mean('nCells')
```
You could write a simple inline function to tie together this redundant code if you wanted.  This could help with readability, e.g., something like

``` python

def sel_data(ds):
     var = ds["timeSeriesStatsMonthly_avg_iceAreaCell_1"]
     return var.where(var > 0).mean('nCells')

sel_data(ds).plot()
sel_data(ds2).plot()
...
```

milenaveneziani commented:
@pwolfram: thanks very much for the feedback. Could you please take a look at the new code? I have addressed most of your comments, except for the last one.. which I just realized now that I have completely missed. Let me get to it.

milenaveneziani commented:
Done. Let me know if any other change would be desirable. Thanks.

pwolfram commented:
@milenaveneziani, I've added some changes via a PR to your branch.  I think we are getting close here.  My plan, to make the transition easier in the future, is to fulfill #7 soon to help streamline deployment of changes to the wrapper in the future.

pwolfram commented:
@milenaveneziani, I'm not sure if the last comment is clear but my plan is to merge this PR before moving onto resolving #7 which should make life easier in the future in terms of installation.
